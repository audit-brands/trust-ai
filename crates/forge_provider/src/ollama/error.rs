use thiserror::Error;

/// Comprehensive error types for Ollama provider operations
#[derive(Debug, Error)]
pub enum OllamaError {
    /// Connection-related errors
    #[error("Failed to connect to Ollama service at {url}: {source}")]
    ConnectionFailed {
        url: String,
        #[source]
        source: reqwest::Error,
    },

    /// Service unavailable errors
    #[error(
        "Ollama service is unavailable. Please ensure Ollama is running and accessible at {url}"
    )]
    ServiceUnavailable { url: String },

    /// Authentication and permission errors
    #[error("Authentication failed for Ollama service: {message}")]
    AuthenticationFailed { message: String },

    /// Model-related errors
    #[error("Model '{model}' not found. Available models can be listed using the models command")]
    ModelNotFound { model: String },

    #[error("Model '{model}' is currently loading. Please wait and try again")]
    ModelLoading { model: String },

    #[error("Model '{model}' failed to load: {reason}")]
    ModelLoadFailed { model: String, reason: String },

    /// Request validation errors
    #[error("Invalid request: {message}")]
    InvalidRequest { message: String },

    #[error("Request payload too large: {size} bytes exceeds maximum allowed")]
    PayloadTooLarge { size: usize },

    #[error("Request timeout after {timeout_seconds} seconds")]
    RequestTimeout { timeout_seconds: u64 },

    /// Response parsing errors
    #[error("Failed to parse response from Ollama: {message}")]
    ResponseParsingFailed { message: String },

    #[error("Received malformed response from Ollama service")]
    MalformedResponse,

    #[error("Unexpected response format from Ollama service")]
    UnexpectedResponseFormat,

    /// Streaming errors
    #[error("Stream interrupted unexpectedly: {reason}")]
    StreamInterrupted { reason: String },

    #[error("Stream parsing failed: {message}")]
    StreamParsingFailed { message: String },

    /// Resource and system errors
    #[error("Insufficient system resources: {message}")]
    InsufficientResources { message: String },

    #[error("Service rate limit exceeded. Please wait before making more requests")]
    RateLimitExceeded,

    /// Configuration errors
    #[error("Invalid Ollama configuration: {message}")]
    InvalidConfiguration { message: String },

    #[error("Ollama base URL is invalid: {url}")]
    InvalidBaseUrl { url: String },

    /// Generic HTTP errors with context
    #[error("HTTP error {status}: {message}")]
    HttpError { status: u16, message: String },

    /// Unknown errors
    #[error("Unknown Ollama error: {message}")]
    Unknown { message: String },
}

impl OllamaError {
    /// Create a connection failed error
    pub fn connection_failed(url: String, source: reqwest::Error) -> Self {
        Self::ConnectionFailed { url, source }
    }

    /// Create a service unavailable error
    pub fn service_unavailable(url: String) -> Self {
        Self::ServiceUnavailable { url }
    }

    /// Create a model not found error
    pub fn model_not_found(model: String) -> Self {
        Self::ModelNotFound { model }
    }

    /// Create a model loading error
    pub fn model_loading(model: String) -> Self {
        Self::ModelLoading { model }
    }

    /// Create a model load failed error
    pub fn model_load_failed(model: String, reason: String) -> Self {
        Self::ModelLoadFailed { model, reason }
    }

    /// Create an invalid request error
    pub fn invalid_request(message: String) -> Self {
        Self::InvalidRequest { message }
    }

    /// Create a response parsing failed error
    pub fn response_parsing_failed(message: String) -> Self {
        Self::ResponseParsingFailed { message }
    }

    /// Create a stream interrupted error
    pub fn stream_interrupted(reason: String) -> Self {
        Self::StreamInterrupted { reason }
    }

    /// Create an HTTP error with status and message
    pub fn http_error(status: u16, message: String) -> Self {
        Self::HttpError { status, message }
    }

    /// Check if this error indicates the service is unavailable
    pub fn is_service_unavailable(&self) -> bool {
        matches!(
            self,
            OllamaError::ServiceUnavailable { .. }
                | OllamaError::ConnectionFailed { .. }
                | OllamaError::HttpError { status: 503, .. }
        )
    }

    /// Check if this error is retryable
    pub fn is_retryable(&self) -> bool {
        matches!(
            self,
            OllamaError::ServiceUnavailable { .. }
                | OllamaError::ConnectionFailed { .. }
                | OllamaError::RequestTimeout { .. }
                | OllamaError::ModelLoading { .. }
                | OllamaError::RateLimitExceeded
                | OllamaError::HttpError { status: 429 | 502 | 503 | 504, .. }
        )
    }

    /// Check if this error indicates a client-side issue
    pub fn is_client_error(&self) -> bool {
        matches!(
            self,
            OllamaError::InvalidRequest { .. }
                | OllamaError::PayloadTooLarge { .. }
                | OllamaError::ModelNotFound { .. }
                | OllamaError::InvalidConfiguration { .. }
                | OllamaError::InvalidBaseUrl { .. }
                | OllamaError::HttpError { status: 400..=499, .. }
        )
    }

    /// Get user-friendly error message with actionable guidance
    pub fn user_message(&self) -> String {
        match self {
            OllamaError::ServiceUnavailable { url } => {
                format!(
                    "Ollama service is not running. Please start Ollama and ensure it's accessible at {url}"
                )
            }
            OllamaError::ConnectionFailed { url, .. } => {
                format!(
                    "Cannot connect to Ollama service at {url}. Please check that Ollama is running and the URL is correct"
                )
            }
            OllamaError::ModelNotFound { model } => {
                format!(
                    "Model '{model}' is not available. Use 'ollama list' to see available models or 'ollama pull {model}' to download it"
                )
            }
            OllamaError::ModelLoading { model } => {
                format!("Model '{model}' is currently loading. Please wait a moment and try again")
            }
            OllamaError::AuthenticationFailed { .. } => {
                "Authentication failed. Please check your Ollama configuration and permissions"
                    .to_string()
            }
            OllamaError::RateLimitExceeded => {
                "Too many requests. Please wait a moment before trying again".to_string()
            }
            OllamaError::RequestTimeout { timeout_seconds } => {
                format!(
                    "Request timed out after {timeout_seconds} seconds. The model might be too large or the system is under heavy load"
                )
            }
            OllamaError::InvalidConfiguration { message } => {
                format!("Configuration error: {message}. Please check your Ollama settings")
            }
            _ => self.to_string(),
        }
    }
}

/// Convert reqwest errors to OllamaError with context
impl From<reqwest::Error> for OllamaError {
    fn from(error: reqwest::Error) -> Self {
        if error.is_timeout() {
            OllamaError::RequestTimeout {
                timeout_seconds: 30, // Default timeout assumption
            }
        } else if error.is_connect() {
            OllamaError::ConnectionFailed {
                url: error.url().map(|u| u.to_string()).unwrap_or_default(),
                source: error,
            }
        } else {
            OllamaError::Unknown { message: error.to_string() }
        }
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn test_service_unavailable_detection() {
        let fixture = OllamaError::service_unavailable("http://localhost:11434".to_string());
        let actual = fixture.is_service_unavailable();
        let expected = true;
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_retryable_error_detection() {
        let fixture = OllamaError::model_loading("llama3.2".to_string());
        let actual = fixture.is_retryable();
        let expected = true;
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_client_error_detection() {
        let fixture = OllamaError::model_not_found("nonexistent".to_string());
        let actual = fixture.is_client_error();
        let expected = true;
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_user_message_for_service_unavailable() {
        let fixture = OllamaError::service_unavailable("http://localhost:11434".to_string());
        let actual = fixture.user_message();
        let expected = "Ollama service is not running. Please start Ollama and ensure it's accessible at http://localhost:11434";
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_user_message_for_model_not_found() {
        let fixture = OllamaError::model_not_found("llama3.2".to_string());
        let actual = fixture.user_message();
        let expected = "Model 'llama3.2' is not available. Use 'ollama list' to see available models or 'ollama pull llama3.2' to download it";
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_http_error_status_categorization() {
        let fixture = OllamaError::http_error(404, "Not Found".to_string());
        let actual = fixture.is_client_error();
        let expected = true;
        assert_eq!(actual, expected);

        let fixture = OllamaError::http_error(503, "Service Unavailable".to_string());
        let actual = fixture.is_service_unavailable();
        let expected = true;
        assert_eq!(actual, expected);
    }
}
