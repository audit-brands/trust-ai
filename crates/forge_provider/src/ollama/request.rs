use derive_setters::Setters;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Default, Setters)]
#[setters(into, strip_option)]
pub struct ChatRequest {
    model: String,
    messages: Vec<Message>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    options: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    keep_alive: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Message {
    pub role: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub images: Option<Vec<String>>,
}

impl TryFrom<forge_app::domain::Context> for ChatRequest {
    type Error = anyhow::Error;

    fn try_from(context: forge_app::domain::Context) -> Result<Self, Self::Error> {
        let messages = context
            .messages
            .into_iter()
            .filter_map(|msg| {
                if let forge_app::domain::ContextMessage::Text(text_msg) = msg {
                    Some(Message {
                        role: match text_msg.role {
                            forge_app::domain::Role::System => "system".to_string(),
                            forge_app::domain::Role::User => "user".to_string(),
                            forge_app::domain::Role::Assistant => "assistant".to_string(),
                        },
                        content: text_msg.content,
                        images: None, // TODO: Handle images when needed
                    })
                } else {
                    None // Skip non-text messages for now
                }
            })
            .collect();

        Ok(ChatRequest {
            model: String::new(), // Will be set by the provider
            messages,
            stream: Some(true), // Default to streaming
            format: None,
            options: None,
            keep_alive: None,
        })
    }
}
