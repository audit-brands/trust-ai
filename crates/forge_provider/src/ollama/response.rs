use forge_app::domain::{ChatCompletionMessage, Content, Model, ModelId};
use serde::Deserialize;

// Response for /api/tags endpoint
#[derive(Deserialize, Debug)]
pub struct ListModelsResponse {
    pub models: Vec<OllamaModel>,
}

#[derive(Deserialize, Debug)]
pub struct OllamaModel {
    pub name: String,
    pub model: String,
    pub modified_at: String,
    pub size: u64,
    pub digest: String,
    pub details: ModelDetails,
}

#[derive(Deserialize, Debug)]
pub struct ModelDetails {
    pub parent_model: String,
    pub format: String,
    pub family: String,
    pub families: Vec<String>,
    pub parameter_size: String,
    pub quantization_level: String,
}

impl From<OllamaModel> for Model {
    fn from(value: OllamaModel) -> Self {
        Self {
            id: ModelId::new(value.name.clone()),
            name: Some(value.name),
            description: Some(format!("Ollama model - {}", value.details.family)),
            context_length: None, // Ollama doesn't provide this in the models response
            tools_supported: Some(true), // Most Ollama models support tools
            supports_parallel_tool_calls: None,
            supports_reasoning: None,
        }
    }
}

// Response for /api/chat endpoint (streaming)
#[derive(Deserialize, Debug)]
pub struct ChatResponse {
    pub model: String,
    pub created_at: String,
    pub message: ChatMessage,
    pub done: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_duration: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub load_duration: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_eval_count: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_eval_duration: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub eval_count: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub eval_duration: Option<u64>,
}

#[derive(Deserialize, Debug)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub images: Option<Vec<String>>,
}

impl TryFrom<ChatResponse> for ChatCompletionMessage {
    type Error = anyhow::Error;

    fn try_from(response: ChatResponse) -> Result<Self, Self::Error> {
        Ok(ChatCompletionMessage {
            content: Some(Content::part(response.message.content)),
            reasoning: None, // Ollama doesn't provide reasoning separately
            reasoning_details: None,
            tool_calls: Vec::new(), // TODO: Handle tool calls when needed
            finish_reason: None,    // TODO: Map Ollama finish reasons
            usage: None,            // TODO: Map usage statistics
        })
    }
}
