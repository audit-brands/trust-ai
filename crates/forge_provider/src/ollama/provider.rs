use anyhow::Context as _;
use derive_builder::Builder;
use forge_app::domain::{ChatCompletionMessage, Context, Model, ModelId, ResultStream};
use reqwest::{Client, Url};
use reqwest_eventsource::{Event, RequestBuilderExt};
use tokio_stream::StreamExt;
use tracing::debug;

use super::error::OllamaError;
use super::request::ChatRequest;
use super::response::{ChatResponse, ListModelsResponse};

use crate::utils::format_http_context;

#[derive(Clone, Builder)]
pub struct Ollama {
    client: Client,
    base_url: Url,
}

impl Ollama {
    pub fn builder() -> OllamaBuilder {
        OllamaBuilder::default()
    }

    fn url(&self, path: &str) -> anyhow::Result<Url> {
        // Validate the path doesn't contain certain patterns
        if path.contains("://") || path.contains("..") {
            anyhow::bail!("Invalid path: Contains forbidden patterns");
        }

        // Remove leading slash to avoid double slashes
        let path = path.trim_start_matches('/');

        self.base_url
            .join(path)
            .with_context(|| format!("Failed to append {} to base URL: {}", path, self.base_url))
    }
}

impl Ollama {
    pub async fn chat(
        &self,
        model: ModelId,
        context: Context,
    ) -> ResultStream<ChatCompletionMessage, anyhow::Error> {
        // Convert context to Ollama chat request
        let request = ChatRequest::try_from(context)?
            .model(model.as_str().to_string())
            .stream(true);

        let url = self.url("api/chat")?;
        debug!(url = %url, model = %model, "Connecting to Ollama");

        let es = self
            .client
            .post(url.clone())
            .json(&request)
            .eventsource()
            .with_context(|| format_http_context(None, "POST", &url))?;

        let url_clone = url.clone();
        let url_clone2 = url.clone();
        let model_clone = model.clone();
        let stream = es
            .take_while(|message| !matches!(message, Err(reqwest_eventsource::Error::StreamEnded)))
            .then(move |event| {
                let url_inner = url_clone.clone();
                let model_inner = model_clone.clone();
                async move {
                match event {
                    Ok(event) => match event {
                        Event::Open => None,
                        Event::Message(event) if ["[DONE]", ""].contains(&event.data.as_str()) => {
                            debug!("Received completion from Ollama");
                            None
                        }
                        Event::Message(message) => Some(
                            serde_json::from_str::<ChatResponse>(&message.data)
                                .map_err(|e| OllamaError::StreamParsingFailed { message: e.to_string() })
                                .with_context(|| "Failed to parse Ollama event")
                                .and_then(|event| {
                                    ChatCompletionMessage::try_from(event).with_context(|| {
                                        format!(
                                            "Failed to create completion message: {}",
                                            message.data
                                        )
                                    })
                                }),
                        ),
                    },
                    Err(error) => match error {
                        reqwest_eventsource::Error::StreamEnded => None,
                        reqwest_eventsource::Error::InvalidStatusCode(_, response) => {
                            let status = response.status();
                            let body_text = response.text().await.ok();
                            
                            // Convert to appropriate OllamaError
                            let ollama_error = match status.as_u16() {
                                404 => OllamaError::model_not_found(model_inner.as_str().to_string()),
                                503 => OllamaError::service_unavailable(url_inner.to_string()),
                                _ => OllamaError::http_error(status.as_u16(), 
                                    body_text.clone().unwrap_or_else(|| "Unknown error".to_string())),
                            };
                            
                            Some(Err(anyhow::anyhow!(ollama_error)).with_context(
                                || match body_text {
                                    Some(body) => {
                                        format!("Invalid status code: {status} Reason: {body}")
                                    }
                                    None => {
                                        format!("Invalid status code: {status} Reason: [Unknown]")
                                    }
                                },
                            ))
                        }
                        reqwest_eventsource::Error::InvalidContentType(_, ref response) => {
                            let status_code = response.status();
                            debug!(response = ?response, "Invalid content type");
                            Some(Err(error).with_context(|| format!("Http Status: {status_code}")))
                        }
                        error => {
                            tracing::error!(error = ?error, "Failed to receive chat completion event");
                            Some(Err(error.into()))
                        }
                    },
                }
                }
            })
            .map(move |response| match response {
                Some(Err(err)) => {
                    Some(Err(err).with_context(|| format_http_context(None, "POST", &url_clone2)))
                }
                _ => response,
            });

        Ok(Box::pin(stream.filter_map(|x| x)))
    }

    pub async fn models(&self) -> anyhow::Result<Vec<Model>> {
        let url = self.url("api/tags")?;
        debug!(url = %url, "Fetching models from Ollama");

        let result = self.client.get(url.clone()).send().await;

        match result {
            Err(error) => {
                tracing::error!(error = ?error, "Failed to fetch models");
                
                // Get status before moving error
                let error_status = error.status();
                let error_msg = error.to_string();
                
                // Convert to OllamaError for better user experience
                let ollama_error = if error.is_timeout() {
                    OllamaError::RequestTimeout { timeout_seconds: 30 }
                } else if error.is_connect() {
                    OllamaError::connection_failed(url.to_string(), error)
                } else {
                    OllamaError::Unknown { message: error_msg }
                };
                
                let ctx_msg = format_http_context(error_status, "GET", &url);
                Err(anyhow::anyhow!(ollama_error))
                    .with_context(|| ctx_msg)
                    .with_context(|| "Failed to fetch models")
            }
            Ok(response) => {
                let status = response.status();
                let ctx_msg = format_http_context(Some(response.status()), "GET", &url);
                
                // Handle different status codes with appropriate errors
                if status == 503 {
                    return Err(anyhow::anyhow!(OllamaError::service_unavailable(url.to_string())))
                        .with_context(|| ctx_msg);
                }
                
                let text = response
                    .text()
                    .await
                    .with_context(|| ctx_msg.clone())
                    .with_context(|| "Failed to decode response into text")?;

                if status.is_success() {
                    let response: ListModelsResponse = serde_json::from_str(&text)
                        .map_err(|e| OllamaError::response_parsing_failed(e.to_string()))
                        .with_context(|| ctx_msg)
                        .with_context(|| "Failed to deserialize models response")?;
                    Ok(response.models.into_iter().map(Into::into).collect())
                } else {
                    // Treat non-200 response as error with appropriate categorization
                    let ollama_error = match status.as_u16() {
                        400..=499 => OllamaError::http_error(status.as_u16(), text),
                        500..=599 => OllamaError::service_unavailable(url.to_string()),
                        _ => OllamaError::http_error(status.as_u16(), text),
                    };
                    
                    Err(anyhow::anyhow!(ollama_error))
                        .with_context(|| ctx_msg)
                        .with_context(|| "Failed to fetch the models")
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use forge_app::domain::ContextMessage;

    use super::*;
    use crate::mock_server::{normalize_ports, MockServer};

    fn create_ollama(base_url: &str) -> anyhow::Result<Ollama> {
        Ok(Ollama::builder()
            .client(Client::new())
            .base_url(Url::parse(base_url)?)
            .build()
            .unwrap())
    }

    fn create_mock_models_response() -> serde_json::Value {
        serde_json::json!({
            "models": [
                {
                    "name": "llama3.2:latest",
                    "model": "llama3.2:latest",
                    "modified_at": "2025-05-04T17:37:44.706015396-07:00",
                    "size": 2019393189u64,
                    "digest": "a80c4f17acd55265feec403c7aef86be0c25983ab279d83f3bcd3abbcb5b8b72",
                    "details": {
                        "parent_model": "",
                        "format": "gguf",
                        "family": "llama",
                        "families": ["llama"],
                        "parameter_size": "3.2B",
                        "quantization_level": "Q4_K_M"
                    }
                },
                {
                    "name": "deepseek-r1:latest",
                    "model": "deepseek-r1:latest",
                    "modified_at": "2025-05-10T08:06:48.639712648-07:00",
                    "size": 4683075271u64,
                    "digest": "0a8c266910232fd3291e71e5ba1e058cc5af9d411192cf88b6d30e92b6e73163",
                    "details": {
                        "parent_model": "",
                        "format": "gguf",
                        "family": "qwen2",
                        "families": ["qwen2"],
                        "parameter_size": "7.6B",
                        "quantization_level": "Q4_K_M"
                    }
                }
            ]
        })
    }

    fn create_error_response(message: &str, code: u16) -> serde_json::Value {
        serde_json::json!({
            "error": {
                "code": code,
                "message": message
            }
        })
    }

    fn create_empty_response() -> serde_json::Value {
        serde_json::json!({
            "models": []
        })
    }

    #[tokio::test]
    async fn test_url_for_models() {
        let ollama = Ollama::builder()
            .client(Client::new())
            .base_url(Url::parse("http://localhost:11434/api/").unwrap())
            .build()
            .unwrap();
        assert_eq!(
            ollama.url("tags").unwrap().as_str(),
            "http://localhost:11434/api/tags"
        );
    }

    #[tokio::test]
    async fn test_request_conversion() {
        let model_id = ModelId::new("llama3.2");
        let context = Context::default()
            .add_message(ContextMessage::system("You're an expert assistant."))
            .add_message(ContextMessage::user(
                "Hello, how are you?",
                model_id.clone().into(),
            ));

        let request = ChatRequest::try_from(context)
            .unwrap()
            .model("llama3.2".to_string())
            .stream(true);

        insta::assert_snapshot!(serde_json::to_string_pretty(&request).unwrap());
    }

    #[tokio::test]
    async fn test_fetch_models_success() -> anyhow::Result<()> {
        let mut fixture = MockServer::new().await;
        let mock = fixture
            .mock_ollama_models(create_mock_models_response(), 200)
            .await;
        let ollama = create_ollama(&fixture.url())?;
        let actual = ollama.models().await?;

        mock.assert_async().await;

        // Verify we got the expected models
        assert_eq!(actual.len(), 2);
        insta::assert_json_snapshot!(actual);
        Ok(())
    }

    #[tokio::test]
    async fn test_fetch_models_http_error_status() -> anyhow::Result<()> {
        let mut fixture = MockServer::new().await;
        let mock = fixture
            .mock_ollama_models(create_error_response("Ollama not running", 503), 503)
            .await;

        let ollama = create_ollama(&fixture.url())?;
        let actual = ollama.models().await;

        mock.assert_async().await;

        // Verify that we got an error
        assert!(actual.is_err());
        insta::assert_snapshot!(normalize_ports(format!("{:#?}", actual.unwrap_err())));
        Ok(())
    }

    #[tokio::test]
    async fn test_fetch_models_empty_response() -> anyhow::Result<()> {
        let mut fixture = MockServer::new().await;
        let mock = fixture
            .mock_ollama_models(create_empty_response(), 200)
            .await;

        let ollama = create_ollama(&fixture.url())?;
        let actual = ollama.models().await?;

        mock.assert_async().await;
        assert!(actual.is_empty());
        Ok(())
    }
}
