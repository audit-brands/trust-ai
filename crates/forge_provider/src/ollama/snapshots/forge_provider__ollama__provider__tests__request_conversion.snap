---
source: crates/forge_provider/src/ollama/provider.rs
expression: "serde_json::to_string_pretty(&request).unwrap()"
---
{
  "model": "llama3.2",
  "messages": [
    {
      "role": "system",
      "content": "You're an expert assistant."
    },
    {
      "role": "user",
      "content": "Hello, how are you?"
    }
  ],
  "stream": true
}
